### ISSUE TITLE
Help!!! SOS, using 2 column as primary key, but just update didn't insert!!!

### ISSUE THREAD

**Issue by luohao123** [simolus3/moor#2412]
this is my table:

```
import 'package:drift/drift.dart';

@DataClassName("DbRoomMember")
class RoomMember extends Table {
  TextColumn get id => text().named("id")();
  TextColumn get lastName => text().named("last_name")();
  TextColumn get firstName => text().named("first_name")();
  TextColumn get roomId => text().named("room_id")();
  @override
  Set<Column> get primaryKey => {id, roomId};
}

```

I have 2 primary key,  but now the whole data inserted is 9, but I have all 18 data, whichh means, some of them are updated but not insert.

<img width="317" alt="image" src="https://user-images.githubusercontent.com/49749220/236476843-4271afda-8529-4953-a874-4dd86cd7f163.png">

<img width="674" alt="image" src="https://user-images.githubusercontent.com/49749220/236476890-4deb2a55-158d-4b9b-92c7-8e8f7328a2a4.png">

it was using the maxium room member num, but not totall.

what am wrong????!!!!!!

**Comment by simolus3** [comment]
It's pretty hard for me to determine what's going wrong with the information you've provided.

- How does the code to create inserts look like? Updates don't do insert by default, you'll have to use [upserts](https://drift.simonbinder.eu/docs/getting-started/writing_queries/#upserts) for that.
- If this is related to the primary key, did you change the primary key after installing the app on your device? Changing tables requires a [migration](https://drift.simonbinder.eu/docs/advanced-features/migrations/).

**Comment by luohao123** [comment]
@simolus3  thank u sir!!!!!!! 

- I just using this to insert:

```
Future<int> addRoomMember(DbRoomMember cts) {
    return into(roomMember).insertOnConflictUpdate(cts);
  }
```

- I reinstalled the app for new sql testing, all behaviour same!!!!

what I want: if keep insert unless id and roomId are same at the same time! but now, it was just update not insert!!!

**Comment by luohao123** [comment]
@simolus3 hello!, please help, did I use wrong `insertOnConflictUpdate`?

**Comment by luohao123** [comment]
@simolus3  I notice something like this:

> Unique constraints and conflict targets
Both insertOnConflictUpdate and onConflict: DoUpdate use an DO UPDATE upsert in sql. This requires us to provide a so-called "conflict target", a set of columns to check for uniqueness violations. By default, drift will use the table's primary key as conflict target. That works in most cases, but if you have custom UNIQUE constraints on some columns, you'll need to use the target parameter on DoUpdate in Dart to include those columns.

But I DONT know how to write exactly, can u teach me?

**Comment by luohao123** [comment]
@simolus3  SOSS!!!

<img width="585" alt="image" src="https://user-images.githubusercontent.com/49749220/236491066-c42f3890-0d1e-4023-a7a5-72cbba6166c8.png">

**Comment by simolus3** [comment]
Ok, `insertOnConflictUpdate` should do the right thing (update if the primary key matches an existing row, insert otherwise). And this works for me:

```dart
import 'package:drift/drift.dart';
import 'package:drift/native.dart';

part 'members.g.dart';

@DataClassName("DbRoomMember")
class RoomMember extends Table {
  TextColumn get id => text().named("id")();
  TextColumn get roomId => text().named("room_id")();
  @override
  Set<Column> get primaryKey => {id, roomId};
}

@DriftDatabase(tables: [RoomMember])
class MyDatabase extends _$MyDatabase {
  MyDatabase(super.e);

  @override
  int get schemaVersion => 1;

  Future<int> addRoomMember(DbRoomMember cts) {
    return into(roomMember).insertOnConflictUpdate(cts);
  }
}

void main() async {
  final database = MyDatabase(NativeDatabase.memory());

  final members = [
    DbRoomMember(id: 'foo', roomId: 'bar'),
    DbRoomMember(id: 'foo', roomId: 'bar2'),
    DbRoomMember(id: 'foo', roomId: 'bar'), // Duplicate
    DbRoomMember(id: 'foo2', roomId: 'bar'),
    DbRoomMember(id: 'foo', roomId: 'bar2'),
  ];

  for (final member in members) {
    await database.addRoomMember(member);
    print(await database.roomMember.all().get());
  }
}
```

This prints

```
[DbRoomMember(id: foo, roomId: bar)]
[DbRoomMember(id: foo, roomId: bar), DbRoomMember(id: foo, roomId: bar2)]
[DbRoomMember(id: foo, roomId: bar), DbRoomMember(id: foo, roomId: bar2)]
[DbRoomMember(id: foo, roomId: bar), DbRoomMember(id: foo, roomId: bar2), DbRoomMember(id: foo2, roomId: bar)]
[DbRoomMember(id: foo, roomId: bar), DbRoomMember(id: foo, roomId: bar2), DbRoomMember(id: foo2, roomId: bar)]
```

which shows that the duplicate row did not get inserted.

Did you re-run the build after adding the primary key? Are you sure there are 18 different pairs of (id, roomId) in your data set that you are inserting? If there duplicates, it's expected that the amount of rows is less than 18 in the end.

**Comment by luohao123** [comment]
@simos that's weired, my room_ids print out are all not same for sure, so at least 5 person must have 4x5 records at least, now I only got 9. Let me try again reinstall the app, re create the table to test it

**Comment by luohao123** [comment]
@simolus3  Hello, I tried again, it still didn't inserts but updates:

```
flutter: [Appdata] add roomMembers for group room: 227e07a8-fd6c-46b6-8546-af20028505c3 猪猪屋 3
flutter:     rm: usrItug3Lj2c5 227e07a8-fd6c-46b6-8546-af20028505c3
flutter:     rm: usrZK8kZTzEHC 227e07a8-fd6c-46b6-8546-af20028505c3
flutter:     rm: usrh8hqEpSWLG 227e07a8-fd6c-46b6-8546-af20028505c3
flutter: [Appdata] add roomMembers for group room: 4540b363-e697-43d1-9548-6ca0175ccb3d 少年班🐻 9
flutter:     rm: usrZK8kZTzEHC 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usrItug3Lj2c5 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usr7zIJPlQoeq 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usra4Q8ClulKA 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usra4Q8ClulKA 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usrsfUEJgX95l 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usrefBILxLv4J 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usrqB4hE5xeYK 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter:     rm: usrcGVE5OLVI5 4540b363-e697-43d1-9548-6ca0175ccb3d
flutter: [Appdata] add roomMembers for group room: 2528c562-3098-407e-8afe-a34d3e02bc19 三胖家族 6
flutter:     rm: usrefBILxLv4J 2528c562-3098-407e-8afe-a34d3e02bc19
flutter:     rm: usrcGVE5OLVI5 2528c562-3098-407e-8afe-a34d3e02bc19
flutter:     rm: usrItug3Lj2c5 2528c562-3098-407e-8afe-a34d3e02bc19
flutter:     rm: usrqB4hE5xeYK 2528c562-3098-407e-8afe-a34d3e02bc19
flutter:     rm: usrsfUEJgX95l 2528c562-3098-407e-8afe-a34d3e02bc19
flutter:     rm: usrZK8kZTzEHC 2528c562-3098-407e-8afe-a34d3e02bc19
flutter: [Appdata] add roomMembers for group room: 843d13d1-0f62-483b-8de6-cee7d127c8af 可以了吗 8
flutter:     rm: usrefBILxLv4J 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrItug3Lj2c5 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrZK8kZTzEHC 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrcGVE5OLVI5 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrsfUEJgX95l 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrefBILxLv4J 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrsfUEJgX95l 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter:     rm: usrqB4hE5xeYK 843d13d1-0f62-483b-8de6-cee7d127c8af
flutter: [Appdata] add roomMembers for group room: 99432280-3b39-4919-91b8-8717aa325f20 ToMars 5
flutter:     rm: usrItug3Lj2c5 99432280-3b39-4919-91b8-8717aa325f20
flutter:     rm: usrZK8kZTzEHC 99432280-3b39-4919-91b8-8717aa325f20
flutter:     rm: usrefBILxLv4J 99432280-3b39-4919-91b8-8717aa325f20
flutter:     rm: usrqB4hE5xeYK 99432280-3b39-4919-91b8-8717aa325f20
flutter:     rm: usrsfUEJgX95l 99432280-3b39-4919-91b8-8717aa325f20
```

these roomIds are not same right? the last column its the data from server, ther users ID are not same as well.

But I got only 9 records .....

Might it was my sql didn't support upsets??? feature?? how should I do?

sqlite3_flutter_libs: ^0.5.13

Am out of mind now what's happening...

**Comment by simolus3** [comment]
Could you enable `logStatements` again and see what actually gets executed? If your sqlite3 didn't support upserts, there would be an exception.

**Comment by luohao123** [comment]
@simolus3 hello, how to enable logStatements? I am newbie in this.

**Comment by luohao123** [comment]
@simolus3 can u share me your test whole file to me you wrote yesterday? see if I can reproduce the minimal on my side

**Comment by luohao123** [comment]
@simolus3 hello, please help, i am totally lost now

**Comment by simolus3** [comment]
Friendly note that tagging me all the time won't make me respond quicker. I'm here to improve drift and solve actual issues, but I'm not your support agent.

> can u share me your test whole file to me you wrote yesterday?

The file I've posted is complete, put it somewhere (say `lib/members.dart`), run `build_runner build` to generate the part file and then run the Dart file with`dart run lib/members.dart`.

> hello, how to enable logStatements? I am newbie in this.

Sorry, I should have mentioned this directly. `logStatements` is a parameter on the `NativeDatabase` constructor. When enabled, drift will log every SQL statement it generates - this should hopefully pin the problem down.

**Comment by luohao123** [comment]
@simolus3  sorry for disturb!! Thanks for your patcient guidance. I am fool to forget to re run build_runner after changed table!! 

Not I have another question, would u help me a little bit? I don't know how to add 2 conditions when leftOuterJoin with drift:

```
flutter drift final query = select(messages)
.join([leftOuterJoin(roomMember, roomMember.id.equalsExp(messages.fromId))])
..where(messages.roomId.equals(roomId))
..orderBy([
OrderingTerm(expression: messages.sendTime, mode: OrderingMode.desc)
]); how to add 2 conditions in leftOutJoin
```

How can I add both roomId and id match in this query? leftOuterJoin with 2 conditions?

**Comment by luohao123** [comment]
Just solved it!
